src/protocol/events.rs

```diff
@@ -0,0 +1,657 @@
+use crate::{
+    binding_utils::{py_to_rs_invitation_status, py_to_rs_realm_role},
+    ids::{self, RealmID, VlobID},
+    invite::InvitationToken,
+    protocol::Reason,
+};
+use libparsec::protocol::authenticated_cmds::{
+    events_listen::{self, APIEvent},
+    events_subscribe,
+};
+use pyo3::{
+    exceptions::{PyAttributeError, PyNotImplementedError},
+    import_exception,
+    prelude::*,
+    pyclass::CompareOp,
+    types::PyBytes,
+    PyObject, PyResult, Python,
+};
+
+use super::gen_rep;
+
+import_exception!(parsec.api.protcol, ProtocolError);
+
+#[pyclass]
+#[derive(PartialEq, Clone)]
+pub(crate) struct EventsListenReq(pub events_listen::Req);
+
+#[pymethods]
+impl EventsListenReq {
+    #[new]
+    fn new(wait: bool) -> PyResult<Self> {
+        Ok(Self(events_listen::Req { wait }))
+    }
+
+    fn __repr__(&self) -> PyResult<String> {
+        Ok(format!("{:?}", self.0))
+    }
+
+    fn dump<'py>(&self, py: Python<'py>) -> PyResult<&'py PyBytes> {
+        Ok(PyBytes::new(
+            py,
+            &self.0.clone().dump().map_err(ProtocolError::new_err)?,
+        ))
+    }
+
+    #[getter]
+    fn wait(&self) -> PyResult<bool> {
+        Ok(self.0.wait)
+    }
+}
+
+gen_rep!(
+    events_listen,
+    EventsListenRep,
+    { .. },
+    [Cancelled, reason: Reason],
+    [NoEvents]
+);
+
+#[pyclass(extends = EventsListenRep)]
+#[derive(Clone)]
+pub(crate) struct EventsListenRepOk;
+
+#[pyclass]
+#[derive(Clone, PartialEq)]
+pub(crate) enum BackendEvent {
+    DeviceCreated,
+    InviteConduiteUpdated,
+    UserCreated,
+    UserRevoked,
+    OrganizationExpired,
+    Pinged,
+    MessageReceived,
+    InviteStatusChanged,
+    RealmMaintenanceStarted,
+    RealmMaintenanceFinished,
+    RealmVlobsUpdated,
+    RealmRolesUpdated,
+    PkiEnrollmentsUpdated,
+}
+
+#[pymethods]
+impl EventsListenRepOk {
+    #[new]
+    fn new(event_data: EventsListenRep) -> PyResult<(Self, EventsListenRep)> {
+        Ok((Self, event_data))
+    }
+
+    fn __repr__(_self: PyRef<'_, Self>) -> String {
+        format!("{:?}", _self.as_ref().0)
+    }
```

Comment:
I don't understand how I am supposed to use the `gen_proto` macro here since this macro expects a `0` attribute on `Self` but `EventsListenRepOk` don't have one. Here the purpose of `EventsListenRepOk` is to wrap any kind of sucessful event command result and use the `Rep` contained in its parent. Is my design erroneous?